### PR TITLE
Introduce foundations for MIF-based decorrelation

### LINKED ISSUE #111: Decorrelation via iterated filters
While working with PRISMA data, it became clear that the decorrelation step using an estimated covariance matrix (via a whitening process) is very heavily dependent on correct selection of that matrix.  Performance of the detectors is then dependent of the accuracy of an initial, largely manual process to "train" the whitening.

There exists an alternative decorrelation process (of which whitening is one example) called Iterative Filtering [1] which decomposes a signal into Intrinsic Mode Functions (oscillating functions around zero) and a trend function (a function with at most one local extremum).  By removing the trend signal, the residual (sum of IMFs) is decorrelated.  The major benefit is that there is no need to estimate a covariance matrix first.  There are multidimensional extensions [2], and examples of use for hyperspectral target detection [3]. 

This issue covers the implementation of the multidimensional iterative filter (MIF).  Subsequent issues will be needed to track the application of MIF to the target detection problem.

References:

[1] Lin, L., Wang, Y., & Zhou, H. (2009). [Iterative filtering as an alternative algorithm for empirical mode decomposition](https://pdfs.semanticscholar.org/fc99/0fda1d6e46f11631f772890bc8da83385cb2.pdf). Advances in Adaptive Data Analysis, 1(04), 543-560.

[2] Cicone, A., & Zhou, H. (2017). [Multidimensional iterative filtering method for the decomposition of high–dimensional non–stationary signals](https://arxiv.org/pdf/1507.07173.pdf). Numerical Mathematics: Theory, Methods and Applications, 10(2), 278-298.

[3] Cicone, A., Liu, J., & Zhou, H. (2016). [Hyperspectral chemical plume detection algorithms based on multidimensional iterative filtering decomposition](https://arxiv.org/pdf/1512.01979.pdf). Philosophical Transactions of the Royal Society A: Mathematical, Physical and Engineering Sciences, 374(2065), 20150196.

### PR DESCRIPTION
**Pull request by jpolchlo**
## Overview
 
This PR provides the means to decorrelate hyperspectral image sources without needing to estimate a covariance matrix.  This removes the need for tricky initial training operations which may not generalize to broader samples drawn from a single sensor's output.

The methodology used here is described in the citations in #111.  Suffice it to say that after transforming an image in this fashion, we should be in position to use a simple spectral angle mapper in place of the normalized matched filter and achieve equivalent or better results, if references are to be believed.

Closes #111 

## Demo

Will provide a notebook containing usage example.

## Notes

So far, this work appears as if it will require GPU acceleration.  Initial CPU-only runs were quite computationally expensive.

## Checklist

- [x] CI passes after rebase
- [x] README.md updated if necessary to reflect the changes
- [x] Jupyter notebook example

### CODE REVIEW

**Comment by jpolchlo** [comment]
Willing to call this PR ready for review.  The code is a bit sloppy and not 100% tested, but that may be OK for now, given that the utility of this approach is minimal.  It does not do what was advertised with respect to decorrelation.  Standard sphering operations are far more effective.  There may be some ability to denoise target detection results by subtracting the first IMF, but this approach was mostly a bust.

**Comment by jamesmcclain** [comment]
:eyes:

File: src/hyperspectral/hyperspectral/decorrelation/mif.py
Comment (jamesmcclain) [inline src/hyperspectral/hyperspectral/decorrelation/mif.py:339]: It looks like you are applying the pre-processing algorithm from the paper cited in the header comment to every band?  I think it would be helpful to add a comment saying this explicitly.

File: src/hyperspectral/hyperspectral/decorrelation/mif.py
Comment (jamesmcclain) [inline src/hyperspectral/hyperspectral/decorrelation/mif.py:340]: Where is `do_mif` defined?

File: src/hyperspectral/hyperspectral/decorrelation/mif.py
Comment (jamesmcclain) [inline src/hyperspectral/hyperspectral/decorrelation/mif.py:342]: I can't find `do_mif` so I don't know exactly what is in `resid[0, 0, :, :]`, but if it is just the last `imf` -- and if results are not satisfactory -- then it might make sense to experiment with subtracting more than just the last one (or changing `max_imfs`, &c).

**Review by jamesmcclain — COMMENTED** [review]
What is here looks okay to me, but I think it might be worth exploring including the MIF-based post-processing algorithm which only needs the first IMF so it should be faster to run(?).

**Review by pomadchin — COMMENTED** [review]
LGTM and once @jamesmcclain gives a 👍 let's merge it 🚀

**Review by jamesmcclain — APPROVED** [review]
:+1: , but it may or may not be worth considering my general comments.

**Review by pomadchin — APPROVED** [review]

**Comment by jamesmcclain** [comment]
:tada:

**Pull request merged**
